monitor/launch/launcher.go

```diff
@@ -96,3 +121,78 @@ func (nm *NodeMonitor) setState(state *models.MonitorStatus) {
 	nm.state = state
 	nm.Mutex.Unlock()
 }
+
+// SendEvent sends an event to the node monitor which will then distribute to
+// all monitor listeners
+func (nm *NodeMonitor) SendEvent(typ int, event interface{}) error {
+	var buf bytes.Buffer
+
+	if err := gob.NewEncoder(&buf).Encode(event); err != nil {
+		nm.bumpLost()
+		return fmt.Errorf("Unable to gob encode: %s", err)
+	}
+
+	select {
+	case nm.queue <- append([]byte{byte(typ)}, buf.Bytes()...):
+	default:
+		nm.bumpLost()
+		return fmt.Errorf("Monitor queue is full, discarding notification")
+	}
+
+	return nil
+}
+
+// bumpLost accounts for a lost notification
+func (nm *NodeMonitor) bumpLost() {
+	nm.pipeLock.Lock()
+	nm.lost++
+	nm.pipeLock.Unlock()
+}
+
+// lostSinceLastTime returns the number of lost samples since the last call to
+// lostSinceLastTime(), the pipeLock must be held for writing.
+func (nm *NodeMonitor) lostSinceLastTime() uint64 {
+	delta := nm.lost - nm.lostLast
+	nm.lostLast = nm.lost
+	return delta
+}
+
+func (nm *NodeMonitor) eventDrainer() {
+	for {
+		if err := nm.send(<-nm.queue); err != nil {
+			log.WithError(err).Warning("Unable to send monitor notification")
+		}
+	}
+}
+
+func (nm *NodeMonitor) send(data []byte) error {
+	nm.pipeLock.Lock()
+	defer nm.pipeLock.Unlock()
+
+	if nm.pipe == nil {
+		return fmt.Errorf("monitor pipe not opened")
+	}
+
+	p := payload.Payload{Data: data, CPU: 0, Lost: nm.lostSinceLastTime(), Type: payload.EventSample}
+
+	payloadBuf, err := p.Encode()
+	if err != nil {
+		return fmt.Errorf("Unable to encode payload: %s", err)
+	}
+
+	meta := &payload.Meta{Size: uint32(len(payloadBuf))}
+	metaBuf, err := meta.MarshalBinary()
+	if err != nil {
+		return fmt.Errorf("Unable to encode metadata: %s", err)
+	}
+
+	msgBuf := append(metaBuf, payloadBuf...)
```

Comment:
Can you add a comment into this code? I'm failing to understand why you are appending `payloadBuf` to `metaBuf`

Reply:
The meta header comes before the payload. I'll move it into a func.

Reply:
Fixed